pkg/ruler/remotequerier.go

```diff
@@ -295,31 +278,31 @@ func (q *RemoteQuerier) createRequest(ctx context.Context, query string, ts time
 	case formatProtobuf:
 		acceptHeader = protobufDecoderInstance.ContentType() + "," + jsonDecoderInstance.ContentType()
 	default:
-		return httpgrpc.HTTPRequest{}, fmt.Errorf("unknown response format '%s'", q.preferredQueryResultResponseFormat)
+		return http.Request{}, fmt.Errorf("unknown response format '%s'", q.preferredQueryResultResponseFormat)
 	}
 
-	req := httpgrpc.HTTPRequest{
+	req := http.Request{
 		Method: http.MethodPost,
-		Url:    q.promHTTPPrefix + queryEndpointPath,
-		Body:   body,
-		Headers: injectHTTPGrpcReadConsistencyHeader(ctx, []*httpgrpc.Header{
-			{Key: textproto.CanonicalMIMEHeaderKey("User-Agent"), Values: []string{version.UserAgent()}},
-			{Key: textproto.CanonicalMIMEHeaderKey("Content-Type"), Values: []string{mimeTypeFormPost}},
-			{Key: textproto.CanonicalMIMEHeaderKey("Content-Length"), Values: []string{strconv.Itoa(len(body))}},
-			{Key: textproto.CanonicalMIMEHeaderKey("Accept"), Values: []string{acceptHeader}},
+		URL:    q.promHTTPUrl.JoinPath(queryEndpointPath),
+		Body:   wrapBuffer(body),
+		Header: injectHTTPGrpcReadConsistencyHeader(ctx, http.Header{
+			textproto.CanonicalMIMEHeaderKey("User-Agent"):     []string{version.UserAgent()},
+			textproto.CanonicalMIMEHeaderKey("Content-Type"):   []string{mimeTypeFormPost},
+			textproto.CanonicalMIMEHeaderKey("Content-Length"): []string{strconv.Itoa(len(body))},
+			textproto.CanonicalMIMEHeaderKey("Accept"):         []string{acceptHeader},
 		}),
 	}
 
 	for _, mdw := range q.middlewares {
 		if err := mdw(ctx, &req); err != nil {
-			return httpgrpc.HTTPRequest{}, err
+			return http.Request{}, err
 		}
 	}
 
 	return req, nil
 }
 
-func (q *RemoteQuerier) sendRequest(ctx context.Context, req *httpgrpc.HTTPRequest, logger log.Logger) (*httpgrpc.HTTPResponse, error) {
+func (q *RemoteQuerier) sendRequest(ctx context.Context, req *http.Request, logger log.Logger) (*http.Response, error) {
```

Comment:
let's drop the `ctx` argument and use `req.Context()` and ensure the request we get here has the correct context. I think this also requires to change up a bit the order of functions that call sendRequest - we should first wrap the context in a timeout, then create a new request which the timeout context, then pass the request to sendRequest